File: leetcode/p0347_top_k_frequent_elements.rs
use std::collections::HashMap;
use std::collections::BinaryHeap;

fn main() {}

struct Solution;

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        if nums.len() <= 1 { return nums }

        let mut map = HashMap::new();
        for num in nums {
            let stat = map.entry(num).or_insert(1);
            *stat += 1;
        }

        let mut heap = BinaryHeap::new();
        for (k, v) in map {
            heap.push((v, k));
        }

        let mut sorted = heap.into_sorted_vec();
        sorted.reverse();

        sorted.iter().take(k as usize).map(|item| item.1).collect()
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test1() {
        assert_eq!(Solution::top_k_frequent(vec![1,1,1,2,2,3], 2), vec![1,2]);
        assert_eq!(Solution::top_k_frequent(vec![1,1,1,2,2,3], 3), vec![1,2,3]);
    }
}

File: leetcode/p0347-top-k-frequent-elements.py
#!/usr/bin/env python3

from typing import List
from collections import defaultdict

class Solution(object):
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        map = defaultdict(int)
        for num in nums:
            map[num] += 1
        bucket = []
        for _ in range(max(map.values()) + 1):
            bucket.append([])
        for (key, val) in map.items():
            bucket[val].append(key)
        res = []
        bucket.reverse()
        for vals in bucket:
            for val in vals:
                res.append(val)
                if len(res) == k:
                    return res

if __name__ == '__main__':
    assert Solution().topKFrequent([1,1,1,2,2,3], 2) == [1,2]
    assert Solution().topKFrequent([1], 1) == [1]